src/python/pants/backend/python/goals/pytest_runner.py

```diff
@@ -96,6 +94,78 @@ def is_conftest_or_type_stub(self) -> bool:
         return file_name.name == "conftest.py" or file_name.suffix == ".pyi"
 
 
+# -----------------------------------------------------------------------------------------
+# Plugin hook
+# -----------------------------------------------------------------------------------------
+
+
+@dataclass(frozen=True)
+class PytestPluginSetup:
+    """The result of custom set up logic before Pytest runs.
+
+    Please reach out it if you would like certain functionality, such as allowing your plugin to set
+    environment variables.
+    """
+
+    digest: Digest = EMPTY_DIGEST
```

Comment:
Adding a digest here is interesting. I was just working on another plugin that used codegen to inject some metadata for tests based on target metadata. Maybe it would be simpler to do that with this.

Reply:
Either approach is valid. It sounds like this logic is purely for test setup though, and not relevant to other goals like `./pants package`. So, I'd bias towards using this plugin hook because it's simpler and the surface area is smaller.

Reply:
For example, looking at your specific implementation, you could get rid of the `relocated_files` hack where you subclass `Sources` and make the field private. Still keep the `stevedore_extension` target, and still keep the `stevedore_namespaces` field on `python_tests` target, but handle the generation of the code with this Pytest plugin hook instead of the more generic codegen hook.